### ISSUE TITLE
Kill client auth spam (NIP-42), or block relays (NIP-65)

### ISSUE THREAD

**Issue by badonyx** [nostr-protocol/nips#818]
Client authentication events (NIP-42) give relays far too much freedom to spam clients. In clients using the gossip model (gossip, snort, iris) I seem to always end up with `wss://eden.nostr.land` in my connected relays, even when I have removed it from my relay list (NIP-65). This relay (and its friends) spams me again and again and again and again to sign client auth events, whether I sign them or not, probably because they are blocking events from users who have not paid. I suspect that this is potentially a major source of issues with client performance and battery drainage, since most users will probably just permit their client to sign all messages to avoid this, or the client will simply do so by default if the key is unlocked.

I propose some possible solutions to this problem. 
1. Add relay blocking to NIP-65 to instruct clients explicitly NOT to connect to certain relays, in case they are misbehaving.
2. Add a clause to NIP-42 that clients SHOULD timeout relays requesting repeat client authentication events in a small window.
3. Add a clause to NIP-42 that relays SHOULD timeout themselves from sending repeat client authentication requests. This is already sort of suggested but it is not stated strongly enough.

cc @Semisol @v0l @mikedilger

**Comment by staab** [comment]
I think this should mostly be solved at the application level. There may be use cases for repeated AUTH (for example authenticating with multiple keys for different uses). I do think a new list for blocked relays might be a good improvement though.

**Comment by arthurfranca** [comment]
#571 NIP-43 fixes it xD~ There are no AUTH messages with NIP-43 ^___^

**Comment by badonyx** [comment]
From nostr:note1e7wkww9r3gwxhr7a8y9lhdrgj2qne8emw2ep3wny6yhlz607xxjq0r2wda

https://media.nostr.build/p/zlEZ.webm

This is the same behavior I'm seeing on snort/iris but on agora/coracle it might happen if you chosen to connect to a misbehaving relay.

**Comment by badonyx** [comment]
> Source for this?
> Even if you got 1 auth request per second, that contributes to 0.1% or less of the resource use of the client. More goes into signature validation, rendering, etc.

Baseless suspicion... maybe worded too strongly, and you are probably right.

**Comment by badonyx** [comment]
@Semisol thanks, relays are behaving better now. On snort/iris I now get one prompt per relay per refresh. I guess that is the expected protocol flow but I feel like it could be better. Maybe for non-sensitive resources the relay could cache authentication by pubkey rather than by connection.. something like that.

**Comment by arthurfranca** [comment]
> I feel like we need a client-initiated AUTH protocol then, or something.

NIP-43 is client-initiated AUTH at connection start. The client would still have to ask NIP-07 extension to sign an event for each new relay connection.

**Comment by arthurfranca** [comment]
oh in this case no. with NIP-43 an account switch would be a reconnect

**Comment by mikedilger** [comment]
I don't think this is a protocol issue, or even a required "norms" isse. I think the protocol (NIP-42 and potentially NIP-43) is okay as it is.  But clients and relays need to deal with each other's worst situations.

I dealt with eden.nostr.land by setting my client to not connect to it (in gossip, set rank=0, even gossip model relay picks won't pick it).  I will probably unban it now as it sounds as though some bug about this has been fixed.

**Comment by badonyx** [comment]
Every time I use Coracle it triggers popups from my signer for client authentication from each of these relays, none of which I have intentionally solicited.

```
eden.nostr.land
atlas.nostr.land
puravida.nostr.land
nostr-pub.semisol.dev
filter.nostr.wine
```

NIP-43 (#571) seems like a good direction toward less obnoxious client auth.

**Comment by staab** [comment]
Yes, because of NIP 65 preferences. Again, all of this should be solved at the application layer. NIP 43 won't change anything regarding relay selection or signing requests.

**Comment by badonyx** [comment]
> NIP 43 won't change anything regarding relay selection or signing requests.

I'm not convinced and have commented on the PR suggesting an improved flow that would help to address this.
